File: BloomFilterTest.java
// javac BloomFilterTest.java
//
// java BloomFilterTest

import java.util.BitSet;

public class BloomFilterTest {
    private BitSet bits;
    private int size;
    private int hashCount;

    // Constructor
    public BloomFilterTest(int m, int k) {
        this.size = m;
        this.hashCount = k;
        this.bits = new BitSet(m);
    }

    // A simple hash function using the item's hashCode and a seed
    private int getHash(String item, int seed) {
        // Combining item and seed to create unique hashes
        int hash = (item + seed).hashCode();
        return Math.abs(hash % size);
    }

    // Add item to filter
    public void add(String item) {
        for (int i = 0; i < hashCount; i++) {
            bits.set(getHash(item, i));
        }
    }

    // Check if item might exist
    public boolean exists(String item) {
        for (int i = 0; i < hashCount; i++) {
            if (!bits.get(getHash(item, i))) {
                return false; // Definitely not there
            }
        }
        return true; // Might be there
    }

    public static void main(String[] args) {
        // Initialize: 100 bits, 3 hash functions
        BloomFilterTest bf = new BloomFilterTest(100, 3);

        // 1. New Testament Seed Items
        String[] additions = { "Fisher-Net", "Alabaster-Jar", "Thorns-Crown", "Seamless-Robe", "Five-Loaves" };

        System.out.println("--- Seeding Bloom Filter ---");
        for (String item : additions) {
            bf.add(item);
            System.out.println("Added: " + item);
        }

        // 2. Testing
        System.out.println("\n--- Testing Membership ---");
        String[] tests = { "Fisher-Net", "Thorns-Crown", "Golden-Calf", "Centurion-Spear", "Fisher-Boat" };

        for (String test : tests) {
            boolean result = bf.exists(test);
            String status = result ? "PROBABLY PRESENT" : "DEFINITELY ABSENT";
            System.out.printf("%-18s : %s%n", test, status);
        }
    }
}

File: BloomFilterTest.rs
// rustc BloomFilterTest.rs
//
// ./BloomFilterTest

struct BloomFilterTest {
    bits: Vec<u8>,
    size: usize,      // Number of bits
    hash_count: u32,
}

impl BloomFilterTest {
    fn new(size: usize, hash_count: u32) -> Self {
        // We need (size/8) bytes to store 'size' bits
        let byte_count = (size + 7) / 8;
        BloomFilterTest {
            bits: vec![0; byte_count],
            size,
            hash_count,
        }
    }

    fn get_hash(&self, item: &str, seed: u32) -> usize {
        // A simple hash: Fowler-Noll-Vo (FNV) style logic
        let mut h: u64 = 0x811c9dc5 ^ (seed as u64);
        for byte in item.bytes() {
            h = h.wrapping_mul(0x01000193);
            h ^= byte as u64;
        }
        (h as usize) % self.size
    }

    fn add(&mut self, item: &str) {
        for i in 0..self.hash_count {
            let idx = self.get_hash(item, i);
            let byte_idx = idx / 8;
            let bit_idx = idx % 8;
            // Set the specific bit to 1 using bitwise OR
            self.bits[byte_idx] |= 1 << bit_idx;
        }
    }

    fn exists(&self, item: &str) -> bool {
        for i in 0..self.hash_count {
            let idx = self.get_hash(item, i);
            let byte_idx = idx / 8;
            let bit_idx = idx % 8;
            // Check if the bit is 0 using bitwise AND
            if (self.bits[byte_idx] & (1 << bit_idx)) == 0 {
                return false; // Definitely not there
            }
        }
        true // Might be there
    }
}

fn main() {
    let mut bf = BloomFilterTest::new(100, 3);

    // Seeding with New Testament items
    let additions = vec!["Fisher-Net", "Alabaster-Jar", "Thorns-Crown", "Seamless-Robe"];
    
    println!("--- Seeding Rust Bloom Filter ---");
    for item in additions {
        bf.add(item);
        println!("Added: {}", item);
    }

    println!("\n--- Testing Membership ---");
    let tests = vec!["Fisher-Net", "Thorns-Crown", "Golden-Calf", "Centurion-Spear", "Fisher-Boat"];

    for test in tests {
        let status = if bf.exists(test) {
            "PROBABLY PRESENT"
        } else {
            "DEFINITELY ABSENT"
        };
        println!("{:<18} : {}", test, status);
    }
}